File: truevfs-extension/truevfs-extension-logging/src/main/java/net/java/truevfs/extension/logging/LogOutputChannel.java
/*
 * Copyright (C) 2005-2012 Schlichtherle IT Services.
 * All rights reserved. Use is subject to license terms.
 */
package net.java.truevfs.extension.logging;

import edu.umd.cs.findbugs.annotations.CreatesObligation;
import java.io.IOException;
import java.nio.channels.SeekableByteChannel;
import javax.annotation.concurrent.Immutable;
import net.java.truecommons.io.DecoratingSeekableChannel;
import net.java.truevfs.kernel.spec.cio.Entry;
import net.java.truevfs.kernel.spec.cio.OutputSocket;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * @author Christian Schlichtherle
 */
@Immutable
final class LogOutputChannel extends DecoratingSeekableChannel {
    private static final Logger
            logger = LoggerFactory.getLogger(LogOutputChannel.class);

    private final OutputSocket<? extends Entry> origin;

    @CreatesObligation
    LogOutputChannel(
            final OutputSocket<? extends Entry> origin,
            final SeekableByteChannel channel) {
        super(channel);
        this.origin = origin;
        log("Opened output channel for {}");
    }

    @Override
    public void close() throws IOException {
        channel.close();
        log("Closed output channel for {}");
    }

    private void log(String message) {
        Entry entry;
        try {
            entry = origin.target();
        } catch (final IOException ignore) {
            entry = null;
        }
        logger.debug(message, entry);
        if (logger.isTraceEnabled())
            logger.trace("Stack trace:", new Throwable());
    }
}

File: truevfs-ext/truevfs-ext-logging/src/main/scala/net/java/truevfs/ext/logging/LogOutputChannel.scala
/*
 * Copyright (C) 2005-2013 Schlichtherle IT Services.
 * All rights reserved. Use is subject to license terms.
 */
package net.java.truevfs.ext.logging

import java.nio.channels._
import net.java.truecommons.io._
import net.java.truecommons.cio._
import org.slf4j._

/**
 * @author Christian Schlichtherle
 */
private final class LogOutputChannel(
  override val origin: OutputSocket[_ <: Entry],
  channel: SeekableByteChannel)
extends DecoratingSeekableChannel(channel) with LogCloseable with Immutable {
  override def logger = LogOutputChannel.logger
}

private object LogOutputChannel {
  private val logger = LoggerFactory.getLogger(classOf[LogOutputChannel])
}
